File: baslerbauer_site/baslerbauer_main/openfarms.py
#!/usr/bin/env python

import requests

# Helper classes to connect to openfarms

OPENFARMS_DOMAIN = 'http://f.datalets.ch'
OPENFARMS_API = OPENFARMS_DOMAIN+'/api/v2/'

def url_with_offset(url):
    """A function which takes an url and returns an url builder. The url builder takes an offset"""
    def func(offset=0):
        if offset == 0:
            return url
        else:
            return url+'?offset={}'.format(offset)
    return func

OPENFARMS_URLS = {
    'farms': url_with_offset(OPENFARMS_API + 'farms/'),
    'produce': url_with_offset(OPENFARMS_API + 'produce/')
}

def get_json(url):
    response = requests.get(url)
    response.raise_for_status()

    return response.json()


def list_items(url_func):
    """url_func is a function which takes an offset and returns an url"""

    data = get_json(url_func(offset=0))

    # Check how many items there are
    total_count = data['meta']['total_count']
    limit = len(data['items'])

    return data['items'] + list_items_paginated(url_func, total_count, offset=limit)


def list_items_paginated(url_func, total_count, offset):
    if total_count <= offset:
        return []

    data = get_json(url_func(offset))
    limit = len(data['items'])

    return data['items'] + list_items_paginated(url_func, total_count, offset=offset+limit)


def list_farms():
    return list_items(OPENFARMS_URLS['farms'])

def list_produce():
    return list_items(OPENFARMS_URLS['produce'])
    

# DEBUG
if __name__=="__main__":
    print(list_farms())
    print(list_produce())

File: basel_farm/app_main/static/js/openfarms.js

top.openfarms_urls = {farms: 'http://f.datalets.ch/api/v2/farms/', produce: 'http://f.datalets.ch/api/v2/produce/'};
top.openfarms = {
  _farms_got_all: false,
  farms: {},
  _produce_got_all: false,
  produce: {},

  listFarms: function(success, fail) {
    if (top.openfarms._farms_got_all) {
      success(top.openfarms.farms);
    } else {
      $.getJSON(top.openfarms_urls.farms, function(data) { 
        top.openfarms.farms = {};
        $.each(data['items'], function(i,farm) {
          top.openfarms.farms[farm['meta']['detail_url']] = farm;
        });
        top.openfarms._farms_got_all = true;
        success(top.openfarms.farms);
      }).fail(fail);
    }
  },


  listProduce: function(success, fail) {
	 console.log("Get produce List");
    if (top.openfarms._produce_got_all) {
      success(top.openfarms.produce);
    } else {
      $.getJSON(top.openfarms_urls.produce, function(data) { 
        top.openfarms.produce = {};
        $.each(data['items'], function(i,produce) {
          top.openfarms.produce[produce['meta']['detail_url']] = produce;
        });
        top.openfarms._produce_got_all = true;
        success(top.openfarms.produce);
      }).fail(fail);
    }
  },

  getFarm: function(url, success, fail) {
    if (top.openfarms.farms[url]) {
      success(top.openfarms.farms[url]);
    } else {
      $.getJSON(url, function(data) { 
        top.openfarms.farms[url] = data; 
        success(data); 
      }).fail(fail);
    }
  },

  getProduce: function(url, success, fail) {
    if (top.openfarms.produce[url]) {
      success(top.openfarms.produce[url]);
    } else {
      $.getJSON(url, function(data) { top.openfarms.produce[url] = data; success(data); }).fail(fail);
    }
  }
};

